From: Boyuan Yang <byang@debian.org>
Date: Fri, 10 Jan 2025 10:42:58 -0500
Subject: Taglib 1.x and 2.x compat fix

Make the source code compatible to both taglib 1.x and 2.x.

Partially cherry-picked from Arch Linux patch
https://gitlab.archlinux.org/archlinux/packaging/packages/vlc/-/commit/40b8a8a1fc24f594a0b828fbde521c644964efaf
---
 modules/meta_engine/taglib.cpp | 37 +++++++++++++++++++++++++++++++++++--
 1 file changed, 35 insertions(+), 2 deletions(-)

diff --git a/modules/meta_engine/taglib.cpp b/modules/meta_engine/taglib.cpp
index 9909a50..96822c8 100644
--- a/modules/meta_engine/taglib.cpp
+++ b/modules/meta_engine/taglib.cpp
@@ -193,7 +193,11 @@ public:
         ByteVector res(length, 0);
         ssize_t i_read = vlc_stream_Read( m_stream, res.data(), length);
         if (i_read < 0)
+#if TAGLIB_VERSION >= VERSION_INT(2, 0, 0)
+            return ByteVector();
+#else
             return ByteVector::null;
+#endif
         else if ((size_t)i_read != length)
             res.resize(i_read);
         return res;
@@ -496,9 +500,15 @@ static void ReadMetaFromASF( ASF::Tag* tag, demux_meta_t* p_demux_meta, vlc_meta
 
 static void ReadMetaFromBasicTag(const Tag* tag, vlc_meta_t *dest)
 {
+#if TAGLIB_VERSION >= VERSION_INT(2, 0, 0)
+#define SET( accessor, meta )                                                  \
+    if( tag->accessor().isEmpty() )              \
+        vlc_meta_Set##meta( dest, tag->accessor().toCString(true) )
+#else
 #define SET( accessor, meta )                                                  \
     if( !tag->accessor().isNull() && !tag->accessor().isEmpty() )              \
         vlc_meta_Set##meta( dest, tag->accessor().toCString(true) )
+#endif
 #define SETINT( accessor, meta )                                               \
     if( tag->accessor() )                                                      \
     {                                                                          \
@@ -837,18 +847,33 @@ static void ReadMetaFromXiph( Ogg::XiphComment* tag, demux_meta_t* p_demux_meta,
 static void ReadMetaFromMP4( MP4::Tag* tag, demux_meta_t *p_demux_meta, vlc_meta_t* p_meta )
 {
     MP4::Item list;
+#if TAGLIB_VERSION >= VERSION_INT(2, 0, 0)
 #define SET( keyName, metaName )                                                             \
-    if( tag->itemListMap().contains(keyName) )                                               \
+    if( tag->itemMap().contains(keyName) )                                               \
+    {                                                                                        \
+        list = tag->itemMap()[keyName];                                                  \
+        vlc_meta_Set##metaName( p_meta, list.toStringList().front().toCString( true ) );     \
+    }
+#define SET_EXTRA( keyName, metaName )                                                   \
+    if( tag->itemMap().contains(keyName) )                                  \
+    {                                                                                \
+        list = tag->itemMap()[keyName];                                     \
+        vlc_meta_AddExtra( p_meta, metaName, list.toStringList().front().toCString( true ) ); \
+    }
+#else
+#define SET( keyName, metaName )                                                             \
+    if( tag->itemMap().contains(keyName) )                                               \
     {                                                                                        \
         list = tag->itemListMap()[keyName];                                                  \
         vlc_meta_Set##metaName( p_meta, list.toStringList().front().toCString( true ) );     \
     }
 #define SET_EXTRA( keyName, metaName )                                                   \
-    if( tag->itemListMap().contains(keyName) )                                  \
+    if( tag->itemMap().contains(keyName) )                                  \
     {                                                                                \
         list = tag->itemListMap()[keyName];                                     \
         vlc_meta_AddExtra( p_meta, metaName, list.toStringList().front().toCString( true ) ); \
     }
+#endif
 
     SET("----:com.apple.iTunes:MusicBrainz Track Id", TrackID );
     SET_EXTRA("----:com.apple.iTunes:MusicBrainz Album Id", VLC_META_EXTRA_MB_ALBUMID );
@@ -856,9 +881,17 @@ static void ReadMetaFromMP4( MP4::Tag* tag, demux_meta_t *p_demux_meta, vlc_meta
 #undef SET
 #undef SET_EXTRA
 
+#if TAGLIB_VERSION >= VERSION_INT(2, 0, 0)
+    if( tag->itemMap().contains("covr") )
+#else
     if( tag->itemListMap().contains("covr") )
+#endif
     {
+#if TAGLIB_VERSION >= VERSION_INT(2, 0, 0)
+        MP4::CoverArtList list = tag->itemMap()["covr"].toCoverArtList();
+#else
         MP4::CoverArtList list = tag->itemListMap()["covr"].toCoverArtList();
+#endif
         const char *psz_format = list[0].format() == MP4::CoverArt::PNG ? "image/png" : "image/jpeg";
 
         msg_Dbg( p_demux_meta, "Found embedded art (%s) is %i bytes",
